subsys/bluetooth/enocean.c

```diff
@@ -0,0 +1,629 @@
+
+	if (has_shortened_name) {
+		if (len == 23 && commissioning) {
+			handle_switch_commissioning(info, buf, true);
+		}
+
+		return;
+	}
+
+	if (!(payload[8] & (BIT_MASK(3) << 5)) && len >= 12 && len <= 16) {
+		handle_switch_data(info, buf, payload);
+		return;
+	}
+
+	/* Ambiguous */
+	struct net_buf_simple_state state;
+	int err;
+
+	if (len == 29 && commissioning) {
+		net_buf_simple_save(buf, &state);
+		err = handle_switch_commissioning(info, buf, false);
+		if (!err) {
+			/* Doing the opposite of the normal error checking
+			 * pattern on purpose here: If the message *wasn't*
+			 * switch commissioning, it could be sensor data, so we
+			 * should keep parsing.
+			 */
+			return;
+		}
+
+		net_buf_simple_restore(buf, &state);
+	}
+
+	handle_sensor_data(info, buf, payload, len + 1 - SIGNATURE_LEN);
+}
+
+static void store_dirty(struct k_work *work)
+{
+	int err;
+
+	for (int i = 0; i < ARRAY_SIZE(devices); ++i) {
+		if (!(devices[i].flags & FLAG_DIRTY)) {
+			continue;
+		}
+
+		char tag[17];
+
+		encode_tag(tag, i, ENTRY_TAG_SEQ);
+		err = settings_save_one(tag, &devices[i].seq,
+					sizeof(devices[i].seq));
+		if (err) {
+			LOG_WRN("#%u err: %d", i, err);
+			break;
+		}
+
+		LOG_DBG("Stored #%u", i);
+
+		devices[i].flags &= ~FLAG_DIRTY;
+	}
+
+	if (err) {
+		schedule_store();
+	}
+}
+
+static int settings_set(const char *key, size_t len, settings_read_cb read_cb,
+			void *cb_arg)
+{
+	struct bt_enocean_device *dev;
+	struct device_entry entry;
+	const char *tag;
+	int size;
+
+	u32_t index = atoi(key);
+
+	if (index >= ARRAY_SIZE(devices)) {
+		return -ENOMEM;
+	}
+
+	dev = &devices[index];
+
+	(void)settings_name_next(key, &tag);
+
+	if (tag[0] == ENTRY_TAG_DEVICE) {
+		size = read_cb(cb_arg, &entry, sizeof(entry));
+		if (size < sizeof(entry)) {
+			return -EINVAL;
+		}
+
+		memcpy(dev->addr, entry.addr, sizeof(dev->addr));
+		memcpy(dev->key, entry.key, sizeof(dev->key));
+		dev->flags |= FLAG_ACTIVE;
+
+		LOG_DBG("Loaded %02x:%02x:%02x:%02x:%02x:%02x", dev->addr[0],
+			dev->addr[1], dev->addr[2], dev->addr[3], dev->addr[4],
+			dev->addr[5]);
+		return 0;
+	}
+
+	if (tag[0] == ENTRY_TAG_SEQ) {
+		size = read_cb(cb_arg, &dev->seq, sizeof(dev->seq));
+		if (size < sizeof(dev->seq)) {
+			return -EINVAL;
+		}
+
+		return 0;
+	}
+
+	return -EINVAL;
+}
+
+static int process_loaded_devs(void)
+{
+	if (!cb.loaded) {
+		return 0;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(devices); ++i) {
+		if (!(devices[i].flags & FLAG_ACTIVE)) {
+			continue;
+		}
+
+		cb.loaded(&devices[i]);
+	}
+
+	return 0;
+}
+
+SETTINGS_STATIC_HANDLER_DEFINE(bt_enocean, "bt/enocean", NULL, settings_set,
+			       process_loaded_devs, NULL);
+
+void bt_enocean_init(const struct bt_enocean_callbacks *callbacks)
+{
+	if (IS_ENABLED(CONFIG_BT_ENOCEAN_STORE_SEQ)) {
+		k_delayed_work_init(&work, store_dirty);
+	}
+
+	cb = *callbacks;
+
+	process_loaded_devs();
+
+	static struct bt_le_scan_cb scan_cb = { .recv = adv_recv };
```

Comment:
Could add const.

Reply:
Nack, the scan_cb contains a linked list node, which is changed at runtime

Reply:
Good point :)